File: split/vpackage.pas
function vpackage(p: halfword; h: scaled; m: smallnumber; l: scaled): halfword;
label
  50, 10;
var
  r: halfword;
  w, d, x: scaled;
  s: scaled;
  g: halfword;
  o: glueord;
begin
  lastbadness := 0;
  r := getnode(7);
  mem[r].hh.b0 := 1;
  mem[r].hh.b1 := 0;
  mem[r + 4].int := 0;
  mem[r + 5].hh.rh := p;
  w := 0;
  {650:}
d := 0;
  x := 0;
  totalstretch[0] := 0;
  totalshrink[0] := 0;
  totalstretch[1] := 0;
  totalshrink[1] := 0;
  totalstretch[2] := 0;
  totalshrink[2] := 0;
  totalstretch[3] := 0;
  totalshrink[3] := 0{:650}
;
  while p <> 0 do{669:}

  begin
    if (p >= himemmin) then
      confusion(854)
    else
      case mem[p].hh.b0 of
        0, 1, 2, 13:{670:}

        begin
          x := x + d + mem[p + 3].int;
          d := mem[p + 2].int;
          if mem[p].hh.b0 >= 2 then
            s := 0
          else
            s := mem[p + 4].int;
          if mem[p + 1].int + s > w then
            w := mem[p + 1].int + s;
        end{:670}
;
        8:{1359:}
{:1359}
;
        10:{671:}

        begin
          x := x + d;
          d := 0;
          g := mem[p + 1].hh.lh;
          x := x + mem[g + 1].int;
          o := mem[g].hh.b0;
          totalstretch[o] := totalstretch[o] + mem[g + 2].int;
          o := mem[g].hh.b1;
          totalshrink[o] := totalshrink[o] + mem[g + 3].int;
          if mem[p].hh.b1 >= 100 then
          begin
            g := mem[p + 1].hh.rh;
            if mem[g + 1].int > w then
              w := mem[g + 1].int;
          end;
        end{:671}
;
        11:
        begin
          x := x + d + mem[p + 1].int;
          d := 0;
        end;
        others: end;
    p := mem[p].hh.rh;
  end{:669}
;
  mem[r + 1].int := w;
  if d > l then
  begin
    x := x + d - l;
    mem[r + 2].int := l;
  end
  else
    mem[r + 2].int := d;
  {672:}
if m = 1 then
    h := x + h;
  mem[r + 3].int := h;
  x := h - x;
  if x = 0 then
  begin
    mem[r + 5].hh.b0 := 0;
    mem[r + 5].hh.b1 := 0;
    mem[r + 6].gr := 0.0;
    goto 10;
  end
  else if x > 0 then{673:}

  begin
    {659:}
if totalstretch[3] <> 0 then
      o := 3
    else if totalstretch[2] <> 0 then
      o := 2
    else if totalstretch[1] <> 0 then
      o :=
        1
    else
      o := 0{:659}
;
    mem[r + 5].hh.b1 := o;
    mem[r + 5].hh.b0 := 1;
    if totalstretch[o] <> 0 then
      mem[r + 6].gr := x / totalstretch[o]
    else
    begin
      mem[
        r + 5].hh.b0 := 0;
      mem[r + 6].gr := 0.0;
    end;
    if o = 0 then
      if mem[r + 5].hh.rh <> 0 then{674:}

      begin
        lastbadness := badness(x, totalstretch[0]);
        if lastbadness > eqtb[5290].int then
        begin
          println;
          if lastbadness > 100 then
            printnl(843)
          else
            printnl(844);
          print(855);
          printint(lastbadness);
          goto 50;
        end;
      end{:674}
;
    goto 10;
  end{:673}

  else{676:}

  begin
    {665:}
if totalshrink[3] <> 0 then
      o := 3
    else if totalshrink[2] <> 0 then
      o := 2
    else if totalshrink[1] <> 0 then
      o := 1
    else
      o :=
        0{:665}
;
    mem[r + 5].hh.b1 := o;
    mem[r + 5].hh.b0 := 2;
    if totalshrink[o] <> 0 then
      mem[r + 6].gr := (-x) / totalshrink[o]
    else
    begin
      mem
        [r + 5].hh.b0 := 0;
      mem[r + 6].gr := 0.0;
    end;
    if (totalshrink[o] < -x) and (o = 0) and (mem[r + 5].hh.rh <> 0) then
    begin
      lastbadness := 1000000;
      mem[r + 6].gr := 1.0;
      {677:}
if (-x - totalshrink[0] > eqtb[5839].int) or (eqtb[5290].int < 100) then
      begin
        println;
        printnl(856);
        printscaled(-x - totalshrink[0]);
        print(857);
        goto 50;
      end{:677}
;
    end
    else if o = 0 then
      if mem[r + 5].hh.rh <> 0 then{678:}

      begin
        lastbadness :=
          badness(-x, totalshrink[0]);
        if lastbadness > eqtb[5290].int then
        begin
          println;
          printnl(858);
          printint(lastbadness);
          goto 50;
        end;
      end{:678}
;
    goto 10;
  end{:676}
{:672}
;
  50:
    {675:}
if outputactive then
      print(846)
    else
    begin
      if packbeginline <> 0 then
      begin
        print(848);
        printint(abs(packbeginline));
        print(849);
      end
      else
        print(850);
      printint(line);
      println;
    end;
  begindiagnostic;
  showbox(r);
  enddiagnostic(true){:675}
;
  10:
    vpackage := r;
end;

File: cpp/vpackage.cpp
#include "vpackage.h"
#include "print.h"
#include "printint.h"
#include "println.h"
#include "printscaled.h"
#include "printnl.h"
#include "shortdisplay.h"
#include "begindiagnostic.h"
#include "showbox.h"
#include "enddiagnostic.h"
#include "confusion.h"
#include "getnode.h"
#include "freenode.h"
#include "badness.h"
/*#include "newrule.h"*/

static void goto50(halfword r)
{
	if (outputactive)
		print(846); //) has occurred while \output is active
	else
	{
		if (packbeginline)
		{
			print(848); //) in alignment at lines 
			printint(abs(packbeginline));
			print(849); //--
		}
		else
			print(850); //) detected at line 
		printint(line);
		println();
	}
	begindiagnostic();
	showbox(r);
	enddiagnostic(true);
}

halfword vpackage(halfword p, scaled h, smallnumber m, scaled l)
{
	lastbadness = 0;
	auto r = getnode(7);
	type(r) = 1;
	subtype(r) = 0;
	shift_amount(r) = 0;
	list_ptr(r) = p;
	scaled w = 0;
	scaled d = 0;
	scaled x = 0;
	totalstretch[0] = 0;
	totalshrink[0] = 0;
	totalstretch[1] = 0;
	totalshrink[1] = 0;
	totalstretch[2] = 0;
	totalshrink[2] = 0;
	totalstretch[3] = 0;
	totalshrink[3] = 0;
	scaled s;
	glueord o;
	while (p)
	{
		if (p >= himemmin)
			confusion(854); //vpack
		else
			switch (type(p))
			{
				case hlist_node:
				case vlist_node:
				case rule_node:
				case unset_node:
					x += d+height(p);
					d = depth(p);
					if (type(p) >= rule_node) // rule_node ou unset_node
						s = 0;
					else
						s = shift_amount(p);
					if (width(p)+s > w)
						w = width(p)+s;
					break;
				case whatsit_node:
					break;
				case glue_node:
					x += d;
					d = 0;
					g = glue_ptr(p);
					x += width(g);
					o = type(g);
					totalstretch[o] += stretch(g);
					o = subtype(g);
					totalshrink[o] += shrink(g);
					if (subtype(p) >= 100)
					{
						g = glue_ptr(p);
						if (width(g) > w)
							w = width(g);
					}
					break;
				case kern_node:
					x += d+width(p);
					d = 0;
			}
		p = link(p);
	}
	width(r) = w;
	if (d > l)
	{
		x += d-l;
		depth(r) = l;
	}
	else
		depth(r) = d;
	if (m == 1)
		h += x;
	height(r) = h;
	x = h-x;
	if (x == 0)
	{
		glue_sign(r) = normal;
		glue_order(r) = 0;
		glue_set(r) = 0.0;
	}
	else 
		if (x > 0)
		{
			glueord o;
			if (totalstretch[3])
				o = 3;
			else 
				if (totalstretch[2])
					o = 2;
				else 
					if (totalstretch[1])
						o = 1;
					else 
						o = 0;
			glue_order(r) = o;
			glue_sign(r) = stretching;
			if (totalstretch[o])
				glue_set(r) = x/totalstretch[o];
			else
			{
				glue_sign(r) = normal;
				glue_set(r) = 0.0;
			}
			if (o == 0)
			if (list_ptr(r))
			{
				lastbadness = badness(x, totalstretch[0]);
				if (lastbadness > int_par(vbadness_code))
				{
					println();
					if (lastbadness > 100)
						printnl(843); //Underfull
					else
						printnl(844); //Loose
					print(855); //\vbox (badness 
					printint(lastbadness);
					goto50(r);
				}
			}
		}
		else
		{
			glueord o;
			if (totalshrink[3])
				o = 3;
			else 
				if (totalshrink[2])
					o = 2;
				else 
					if (totalshrink[1])
						o = 1;
					else
						o = 0;
			glue_order(r) = o;
			glue_sign(r) = shrinking;
			if (totalshrink[o])
				glue_set(r) = (-x)/totalshrink[o];
			else
			{
				glue_sign(r) = normal;
				glue_set(r) = 0.0;
			}
			if (totalshrink[o] < -x && o == 0 && list_ptr(r))
			{
				lastbadness = 1000000;
				glue_set(r) = 1.0;
				if (-x-totalshrink[0] > dimen_par(vfuzz_code) || int_par(vbadness_code) < 100)
				{
					println();
					printnl(856); //6Overfull \vbox (
					printscaled(-x-totalshrink[0]);
					print(857); //pt too high
					goto50(r);
				}
			}
			else 
				if (o == 0 && list_ptr(r))
				{
					lastbadness = badness(-x, totalshrink[0]);
					if (lastbadness > int_par(vbadness_code))
					{
						println();
						printnl(858); //Tight \vbox (badness 
						printint(lastbadness);
						goto50(r);
					}
				}
		}
	return r;
}
